### ISSUE TITLE
apply_along_axis cuts strings

### ISSUE THREAD

**Issue by lukovnikov** [numpy/numpy#8352]
I'm trying to concatenate all elements of a row into a string as follows:
    
    np.apply_along_axis(lambda x: " ".join(map(str, x)), 1, b)

*b* has a specific value

    [[111,111,0,0,0], [111,111,111,111,111]]

However, the result of the line is:

    ['111 111 0 0 0', '111 111 111 1']

It looks like np.apply_along_axis is cutting the second string to be of the same length as the first one. If I put a longer sequence first, the result is correct:

    ['111 111 111 111 111', '111 111 0 0 0']

 So I'm guessing this is a bug?

**Comment by mhvk** [comment]
(found this via #8363) @lukovnikov - the code fails because it was written with numerical arrays in mind, for which the computation on any part of an array can be expected to return the same type of output as any other part.  I should note more generally that numpy arrays are not particularly good or efficient at strings, and unless you have a very complicated array, my guess is that you would be much better off just working with lists and python functions, especially since you already are using the python string functions to do the concatenating.

**Comment by gfyoung** [comment]
@mhvk : I disagree completely with your statement.  `numpy` might be written for numerical computations, but that doesn't mean we have omit functionality with `str` arrays.  After all, that is why we have specific dtypes for strings unlike libraries like `pandas`.

This is a bug, and it should be patched unless you can come up with a more convincing argument than what you have provided.

**Comment by mhvk** [comment]
@gfyoung - I'm not saying one should not try to solve the bug (though I think it is obvious any solution better not cause a huge performance regression for more typical), just explaining why the bug exists and suggesting that for strings one really is better off not using `ndarray`. Anyway, those are my 2¢.

**Comment by gfyoung** [comment]
@mhvk : Fair enough, though your response came across as if this wasn't really a concern of `numpy`.

**Comment by yvan** [comment]
i had the same issue:

you can see it crops the 'g' of jpg by simply referencing it. i figured it has something to do with the shape change. ended up using lists + map instead.

<img width="1102" alt="capture d ecran 2017-05-27 a 11 54 25" src="https://cloud.githubusercontent.com/assets/4513316/26522669/52086186-42d3-11e7-91bc-fa2b08551efd.png">

**Comment by eric-wieser** [comment]
I'm assuming that that's a deliberately contrived example, because you shouldn't be using `apply_along_axis` for simple indexing like that.

`np.ma.apply_along_axis` will work correctly here (for now - see #8511). Another option (crashed until 1.13) is a manual cast to `dtype` object:

```
np.apply_along_axis(lambda x: np.array(x[0], object), 1, fnames)
```

**Comment by linwoodc3** [comment]
@eric-wieser  your comment above helped me solve a problem I've been having for a few months with numpy and string operations.  👍 

> I'm assuming that that's a deliberately contrived example, because you shouldn't be using apply_along_axis for simple indexing like that.
> 
> np.ma.apply_along_axis will work correctly here (for now - see #8511). Another option (crashed until 1.13) is a manual cast to dtype object:
> 
> np.apply_along_axis(lambda x: np.array(x[0], object), 1, fnames)

Thanks

**Comment by eric-wieser** [comment]
Perhaps we should just add a `dtype=` argument to `apply_along_axis`

**Comment by linwoodc3** [comment]
@eric-wieser 

I agree you should add that argument, if there are performance differences between the mask array version and the regular version.  I just did a small test (not sure if it means anything) and here is a pic of the results:


![screen shot 2017-12-29 at 9 58 56 am](https://user-images.githubusercontent.com/2884914/34439888-3a300144-ec7f-11e7-9edf-6e43a91c8184.png)

**My use case**: I'm storing parsed NLP data (strings) in numpy arrays and trying to get rid of all `for loops` and `if-else` clauses; I apply some text analytics functions using the `apply_along_axis`.  Any speed benefit would be awesome as each information extraction could have 10-20 variations (from a document that could have tens to hundreds of information extractions).

**Comment by eric-wieser** [comment]
> and trying to get rid of all for loops

Note that `apply_along_axis` is just a python for loop, with a little help in allocating the output array for you - it's not unlikely that it's slower than the loop it replaces

**Comment by linwoodc3** [comment]
Ah shucks

**Comment by cerlymarco** [comment]
To avoid the cut when joining string with np.apply_along_axis:

`a = np.array(['sssssssss','ffffffffffffff'])
b = np.array(['cccccccccccc','iiiiiiiiiiiiiiiii'])

def join_txt(text): return np.asarray(" ".join(text),dtype=object)

np.apply_along_axis(join_txt,0,[a,b])`

the result is

`array(['sssssssss cccccccccccc', 'ffffffffffffff iiiiiiiiiiiiiiiii'], dtype=object)`

**Comment by alexcoca** [comment]
@cerlymarco , how does the approach you propose compared to @linwoodc3's solution in terms of speed?

**Comment by eric-wieser** [comment]
> Perhaps we should just add a `dtype=` argument to `apply_along_axis`

Unfortunately this isn't possible without breaking someone. The current signature is:
```
def apply_along_axis(func1d, axis, arr, *args, **kwargs):
```
Today, users can call it as both:
```
def f1(x):
    return x
np.apply_along_axis(f1, 0, my_arr)

def f2(x, *, dtype):
    return x.astype(dtype)
np.apply_along_axis(f2, 0, my_arr, dtype=int)
```
If we make a `apply_along_axis` take a dtype argument and not pass it on to `f`, then `f2` will fail. If we make it take a dtype argument and pass it on to `f`, then `f1` will fail.

What we could do is:
* Emit a `FutureWarning` if `'dtype' in kwargs` telling people to rename their arguments
* Wait 2 years
* Break any users still using something like `f2` above

**Comment by jcpayne** [comment]
Just a comment that I was trying to add a prefix and suffix to a filename using Numpy on a Pandas Series and ran into the same problem (I think); i.e., that the second string was cut to the same length as the first.
```filenames = Series(['S1/C03/C03_R1/S1_C03_R1_PICT0239.JPG','S1/C03/C03_R1/S1_C03_R1_PICT0239.JPG'])
prefix = 'somepath'
np.char.add(prefix, filenames.astype(str))
```
